File: python/street-address.py
import json
import pprint
import urllib

# Obtain an authentication ID/token pair from your
# SmartyStreets account and put them in below.

LOCATION = "https://api.smartystreets.com/street-address/"
QUERY_STRING = urllib.urlencode({   # entire query sting must be URL-Encoded
    "auth-id": r"<ID VALUE>",
    "auth-token": r"<RAW TOKEN VALUE>",
    "street": "1 infinite loop",
    "city": "cupertino",
    "state": "ca",
    "zipcode": "95014",
    "candidates": "1",
})
URL = LOCATION + "?" + QUERY_STRING

response = urllib.urlopen(URL).read()
structure = json.loads(response)
pprint.pprint(structure)

# SAMPLE OUTPUT:
#
#[{u'analysis': {u'dpv_cmra': u'N',
#                u'dpv_footnotes': u'AABB',
#                u'dpv_match_code': u'Y',
#                u'dpv_vacant': u'N',
#                u'ews_match': False},
#  u'candidate_index': 0,
#  u'components': {u'city_name': u'Cupertino',
#                  u'delivery_point': u'01',
#                  u'delivery_point_check_digit': u'7',
#                  u'plus4_code': u'2083',
#                  u'primary_number': u'1',
#                  u'state_abbreviation': u'CA',
#                  u'street_name': u'Infinite',
#                  u'street_suffix': u'Loop',
#                  u'zipcode': u'95014'},
#  u'delivery_line_1': u'1 Infinite Loop',
#  u'delivery_point_barcode': u'950142083017',
#  u'input_index': 0,
#  u'last_line': u'Cupertino CA 95014-2083',
#  u'metadata': {u'carrier_route': u'C067',
#                u'congressional_district': u'15',
#                u'county_fips': u'06085',
#                u'county_name': u'Santa Clara',
#                u'latitude': 37.33118,
#                u'longitude': -122.03062,
#                u'precision': u'Zip9',
#                u'record_type': u'S'}}]

File: ruby/street_address.rb
require 'cgi'
require 'json'
require 'net/http'

# Remember to URL encode anything that could have 
# spaces (street, city) or special characters (auth-token).

# Obtain an authentication ID/token pair from your
# SmartyStreets account and put them in below.

URL = 'api.smartystreets.com'
STREET = CGI::escape('1 infinite loop')
CITY = CGI::escape('cupertino')
STATE = 'ca'
ZIP_CODE = '95014'
NUMBER_OF_CANDIDATES = '1'
AUTH_ID = CGI::escape('YOUR_AUTHENTICATION_ID_HERE')
AUTH_TOKEN = CGI::escape('YOUR_AUTHENTICATION_TOKEN_HERE')

QUERY_STRING = '/street-address/?' +
 	'street=' + STREET +
	'&city=' + CITY +
	'&state=' + STATE +
	'&zipcode=' + ZIP_CODE +
	'&candidates=' + NUMBER_OF_CANDIDATES +
	'&auth-id=' + AUTH_ID +
	'&auth-token=' + AUTH_TOKEN

http = Net::HTTP.new(URL)
request = Net::HTTP::Get.new(QUERY_STRING)
response = http.request(request)
puts JSON.pretty_generate(JSON.parse(response.body))

# SAMPLE OUTPUT:
#
#[
#  {
#    "input_index": 0,
#    "candidate_index": 0,
#    "delivery_line_1": "1 Infinite Loop",
#    "last_line": "Cupertino CA 95014-2083",
#    "delivery_point_barcode": "950142083017",
#    "components": {
#      "primary_number": "1",
#      "street_name": "Infinite",
#      "street_suffix": "Loop",
#      "city_name": "Cupertino",
#      "state_abbreviation": "CA",
#      "zipcode": "95014",
#      "plus4_code": "2083",
#      "delivery_point": "01",
#      "delivery_point_check_digit": "7"
#    },
#    "metadata": {
#      "record_type": "S",
#      "county_fips": "06085",
#      "county_name": "Santa Clara",
#      "carrier_route": "C067",
#      "congressional_district": "15",
#      "latitude": 37.33118,
#      "longitude": -122.03062,
#      "precision": "Zip9"
#    },
#    "analysis": {
#      "dpv_match_code": "Y",
#      "dpv_footnotes": "AABB",
#      "dpv_cmra": "N",
#      "dpv_vacant": "N",
#      "ews_match": false
#    }
#  }
#]